File: benchmark/c/opl/opl_emu_opl_compute_phase_step/opl_emu_opl_compute_phase_step.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

uint32_t opl_emu_bitfield(uint32_t value, int start, int length );
static uint32_t opl_emu_opl_compute_phase_step(uint32_t block_freq, uint32_t multiple, int32_t lfo_raw_pm);

uint32_t opl_emu_bitfield(uint32_t value, int start, int length )
{
	return (value >> start) & ((1 << length) - 1);
}
static uint32_t opl_emu_opl_compute_phase_step(uint32_t block_freq, uint32_t multiple, int32_t lfo_raw_pm)
{
	uint32_t fnum = opl_emu_bitfield(block_freq, 0, 10) << 2;
	fnum += (lfo_raw_pm * opl_emu_bitfield(block_freq, 7, 3)) >> 1;
	fnum &= 0xfff;
	uint32_t block = opl_emu_bitfield(block_freq, 10, 3);
	uint32_t phase_step = (fnum << block) >> 2;
	return (phase_step * multiple) >> 1;
}

File: flourine_reslut/success/c/opl/opl_emu_opl_compute_phase_step/gpt4/BaseRepair/opl_emu_opl_compute_phase_step.rs

fn opl_emu_bitfield(value: u32, start: u32, length: u32) -> u32 {
    (value >> start) & ((1 << length) - 1)
}

fn opl_emu_opl_compute_phase_step(block_freq: u32, multiple: u32, lfo_raw_pm: i32) -> u32 {
    let fnum = opl_emu_bitfield(block_freq, 0, 10) << 2;
    let fnum = fnum.wrapping_add(((lfo_raw_pm as u32) * opl_emu_bitfield(block_freq, 7, 3)) >> 1) & 0xfff;
    let block = opl_emu_bitfield(block_freq, 10, 3);
    let phase_step = (fnum << block) >> 2;
    (phase_step * multiple) >> 1
}
